File: server/tools/embed_lore.py
# Usage:
#   pip install sentence-transformers numpy
#   python embed_lore.py path\to\lore.txt  (writes ../src/rag/lore.json)
import sys, os, json
from sentence_transformers import SentenceTransformer


def chunk_text(text, chunk_words=600):
    words = text.split()
    for i in range(0, len(words), chunk_words):
        yield " ".join(words[i:i+chunk_words])


def main():
    if len(sys.argv) < 2:
        print("Usage: python embed_lore.py <lore_text_file>")
        sys.exit(1)
    src = sys.argv[1]
    with open(src, 'r', encoding='utf-8') as f:
        raw = f.read()

    chunks = [c for c in chunk_text(raw, 600) if c.strip()]
    print(f"[embed] chunks: {len(chunks)}")
    model = SentenceTransformer('all-MiniLM-L6-v2')
    vecs = model.encode(chunks, show_progress_bar=True)

    out = []
    for i, (txt, emb) in enumerate(zip(chunks, vecs)):
        out.append({
            "id": f"chunk_{i}",
            "text": txt,
            "embedding": [float(x) for x in emb.tolist()]
        })

    out_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'src', 'rag'))
    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, 'lore.json')
    with open(out_path, 'w', encoding='utf-8') as f:
        json.dump(out, f, ensure_ascii=False)
    print(f"[embed] wrote {out_path}")


if __name__ == "__main__":
    main()

File: server/tools/embed_lore.js
#!/usr/bin/env node
/**
 * Embed lore from a raw .txt file into JSON vectors using local Transformers.js (no Python).
 * Usage (from repo root):
 *   node server/tools/embed_lore.js path\to\newLore.txt
 * Writes to: server/src/rag/lore.json
 */
const fs = require('fs');
const path = require('path');

const { embedTexts } = require('../src/rag/embeddings_local');

function chunkText(text, chunkWords = 600) {
  const words = String(text || '').split(/\s+/).filter(Boolean);
  const chunks = [];
  for (let i = 0; i < words.length; i += chunkWords) {
    chunks.push(words.slice(i, i + chunkWords).join(' '));
  }
  return chunks.filter(c => c.trim());
}

async function main() {
  const src = process.argv[2];
  if (!src) {
    console.error('Usage: node server/tools/embed_lore.js <lore_text_file>');
    process.exit(1);
  }
  const absSrc = path.resolve(process.cwd(), src);
  if (!fs.existsSync(absSrc)) {
    console.error(`File not found: ${absSrc}`);
    process.exit(1);
  }
  const raw = fs.readFileSync(absSrc, 'utf8');
  const chunks = chunkText(raw, 600);
  console.log(`[embed-js] chunks: ${chunks.length}`);

  const vectors = await embedTexts(chunks);
  if (!vectors || !Array.isArray(vectors) || vectors.length !== chunks.length) {
    console.error('[embed-js] Failed to generate embeddings');
    process.exit(2);
  }

  const out = chunks.map((txt, i) => ({
    id: `chunk_${i}`,
    text: txt,
    embedding: vectors[i].map((x) => Number(x))
  }));

  const outDir = path.resolve(__dirname, '..', 'src', 'rag');
  fs.mkdirSync(outDir, { recursive: true });
  const outPath = path.join(outDir, 'lore.json');
  fs.writeFileSync(outPath, JSON.stringify(out, null, 0), 'utf8');
  console.log(`[embed-js] wrote ${outPath}`);
}

main().catch((e) => {
  console.error('[embed-js] Error:', e?.message || e);
  process.exit(3);
});
